llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp

```diff
@@ -1467,231 +1469,246 @@ void GCNSchedStage::revertScheduling() {
-  // Temporary copies of cached variables we will be modifying and replacing if
-  // sinking succeeds.
-  SmallVector<
-      std::pair<MachineBasicBlock::iterator, MachineBasicBlock::iterator>, 32>
-      NewRegions;
-  DenseMap<unsigned, GCNRPTracker::LiveRegSet> NewLiveIns;
-  DenseMap<unsigned, GCNRegPressure> NewPressure;
-  BitVector NewRescheduleRegions;
-  LiveIntervals *LIS = DAG.LIS;
-
-  NewRegions.resize(DAG.Regions.size());
-  NewRescheduleRegions.resize(DAG.Regions.size());
-
-  // Collect only regions that has a rematerializable def as a live-in.
-  SmallSet<unsigned, 16> ImpactedRegions;
-  for (const auto &It : RematDefToLiveInRegions)
-    ImpactedRegions.insert(It.second.begin(), It.second.end());
-
-  // Make copies of register pressure and live-ins cache that will be updated
-  // as we rematerialize.
-  for (auto Idx : ImpactedRegions) {
-    NewPressure[Idx] = DAG.Pressure[Idx];
-    NewLiveIns[Idx] = DAG.LiveIns[Idx];
+    unsigned NumVGPRs = RP.getVGPRNum(ST.hasGFX90AInsts());
+    unsigned NumToIncreaseOcc = ST.getNumVGPRsToIncreaseOccupancy(NumVGPRs);
+    OptRegions.insert({I, NumToIncreaseOcc});
+    RA_DEBUG(dbgs() << "Region " << I << " has min. occupancy: decrease by "
+                    << NumToIncreaseOcc << " VGPR(s) to improve occupancy\n");
   }
-  NewRegions = DAG.Regions;
-  NewRescheduleRegions.reset();
+  if (OptRegions.empty())
+    return false;
 
-  DenseMap<MachineInstr *, MachineInstr *> InsertedMIToOldDef;
-  bool Improved = false;
-  for (auto I : ImpactedRegions) {
-    if (!DAG.RegionsWithMinOcc[I])
-      continue;
+  // Tracks estimated rematerialization gains (i.e., reduction in RP) for
+  // this instruction in each optimizable region.
+  auto ReduceRPInRegion = [&](auto OptIt, unsigned I,
+                              LaneBitmask Mask) -> bool {
+    auto NumRegs = SIRegisterInfo::getNumCoveredRegs(Mask);
+    unsigned &RPExcess = OptIt->getSecond();
+    if (NumRegs >= RPExcess) {
+      OptRegions.erase(I);
+      LLVM_DEBUG(dbgs() << "sinking increases occupancy in region " << I
+                        << "\n");
+    } else {
+      RPExcess -= NumRegs;
+      LLVM_DEBUG(dbgs() << "sinking reduces excess pressure in region " << I
+                        << " by " << NumRegs << " (" << RPExcess << " left)\n");
+    }
+    return OptRegions.empty();
+  };
+
+  // We need up-to-date live-out info. to query live-out register masks in
+  // regions containing rematerializable instructions.
+  DAG.RegionLiveOuts.buildLiveRegMap();
+
+  for (unsigned I = 0, E = DAG.Regions.size(); I != E; ++I) {
+    auto Region = DAG.Regions[I];
+    for (auto MI = Region.first; MI != Region.second; ++MI) {
+      // We only support instructions with at least one register (but
+      // non-subregister) operand.
+      MachineInstr &DefMI = *MI;
+      if (DefMI.isBundle() || !DefMI.getNumOperands() ||
+          !DefMI.getOperand(0).isReg() || DefMI.getOperand(0).getSubReg())
```

Comment:
Rematerializing undef subregister defs should work. Also does this need to verify that operand 0 is a def? But this is repeating isReallyTriviallyReMaterializeable's logic anyway, most of these pre-checks should be redundant

Reply:
Thanks for the catches. We indeed don't need to check that operand 0 is a def here as it is already implied by `isTriviallyReMaterializable`. I now allow undef subregister defs as well (see `test_occ_7_sink_one_def_of_undef_subreg_for_8`).

However, I am now forced to do something that seems slightly weird to me during rematerialization; I have to explicitly set the subreg on the new MI's def after calling `TII.reMaterialize` even though the subreg is passed to the method already (if I don't do this I hit an assert in `DAG.LIS->createAndComputeVirtRegInterval(Reg)` at the end). Not sure this is the correct way to do this.

Reply:
The subregister should survive the clone and substitute. What assert?

Reply:
[This one in `LiveInterval::computeSubRangeUndefs`](https://github.com/llvm/llvm-project/blob/00b07aa399b870d4705c4fb628e25d62c8ccc27b/llvm/lib/CodeGen/LiveInterval.cpp#L971). Looking at the in-flight MIR before explicitly setting the subreg after rematerialization it seems that the `NewMI`'s def reg is being assigned a different non-intersecting subregister than the original one, which leads to the two masks tested here to be completely unrelated.

In the `test_occ_7_sink_one_def_of_undef_subreg_for_8` test for example, after calling `SIInstrInfo::reMaterialize` but before setting the subreg explicitly, the new MI uses `sub2` instead of `sub1`.

```
[PreRARemat] Rematerializing undef %32.sub1:vreg_64_align2 = V_MOV_B32_e32 23, implicit $exec
[PreRARemat] New MI:         undef %32.sub2:vreg_64_align2 = V_MOV_B32_e32 23, implicit $exec
```